src/Sarif.Converters/SemmleConverter.cs

```diff
@@ -11,8 +15,189 @@ namespace Microsoft.CodeAnalysis.Sarif.Converters
     /// </summary>
     internal class SemmleConverter : ToolFileConverterBase
     {
+        // Semmle logs are CSV files.
+        private static readonly string[] s_delimiters = new[] { "," };
+
+        // The fields are as follows:
+        private enum FieldIndex
+        {
+            QueryName,
+            QueryDescription,
+            Severity,
+            Message,
+            RelativePath,
+            Path,
+            StartLine,
+            StartColumn,
+            EndLine,
+            EndColumn
+        }
+
+        private TextFieldParser _parser;
+        private List<Notification> _toolNotifications;
+
+        /// <summary>
+        /// Converts a Semmle log file in CSV format to a SARIF log file.
+        /// </summary>
+        /// <param name="input">
+        /// Input stream from which to read the Semmle log.
+        /// </param>
+        /// <param name="output">
+        /// Output string to which to write the SARIF log.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown when one or more required arguments are null.
+        /// </exception>
         public override void Convert(Stream input, IResultLogWriter output)
         {
+            if (input == null)
+            {
+                throw new ArgumentNullException(nameof(input));
+            }
+
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(output));
+            }
+
+            _toolNotifications = new List<Notification>();
+
+            var results = GetResultsFromStream(input);
+
+            var tool = new Tool
+            {
+                Name = "Semmle"
+            };
+
+            output.Initialize(id: null, correlationId: null);
+
+            output.WriteTool(tool);
+
+            output.OpenResults();
+            output.WriteResults(results);
+            output.CloseResults();
+
+            if (_toolNotifications.Any())
+            {
+                output.WriteToolNotifications(_toolNotifications);
+            }
+        }
+
+        private Result[] GetResultsFromStream(Stream input)
+        {
+            var results = new List<Result>();
+            using (_parser = new TextFieldParser(input)
+            {
+                TextFieldType = FieldType.Delimited,
+                Delimiters = s_delimiters
+            })
+            {
+                while (!_parser.EndOfData)
+                {
+                    results.Add(ParseResult());
+                }
+            }
+
+            return results.ToArray();
+        }
+
+        private Result ParseResult()
+        {
+            string[] fields = _parser.ReadFields();
+            return new Result
+            {
+                Level = ResultLevelFromSemmleSeverity(GetString(fields, FieldIndex.Severity)),
+                Message = fields[(int)FieldIndex.Message],
```

Comment:
I looked for and didn't find a work item for message formatting, either in the sarif-sdk repo or in the sarif-spec repo. I filed sarif-standard/sarif-spec#275, "Should we allow formatting in messages?".

But the issue of parsing out links from the Semmle output, creating related locations for them, and scrubbing them, is a separate issue, for two reasons:

1. Semmle's embedded link format is unlikely to match whatever format(s) we decide to support for formatted messages:

```
Incorrect lazy initialization of static field [["SKIP_BYTE_BUFFER"|"relative://src/main/java/org/apache/commons/io/IOUtils.java:170:6561:16"]]: Missing synchronization.
```

2. In any case, the act of creating related locations and scrubbing these escape sequences is a separate work item. I filed #546, "Semmle: Create related locations from embedded links in messages".